File: 784_letter_case_permutation/solution.py
def letterCasePermutation(s):
    out = []
    current = []
    A_ID = ord("A")
    z_ID = ord("z")

    def recurse(i):
        if i == len(s):
            new_string = "".join(current)
            out.append(new_string)
        elif A_ID <= ord(s[i]) <= z_ID:
            current.append(s[i].lower())
            recurse(i + 1)
            current.pop()

            current.append(s[i].upper())
            recurse(i + 1)
            current.pop()
        else:
            current.append(s[i])
            recurse(i + 1)
            current.pop()
    recurse(0)
    return out


s = input("input a string: ")
print(letterCasePermutation(s))

File: 784_letter_case_permutation/solution.js
function letterCasePermutation(s) {
  const strings = [];

  const currentWord = [];
  function recurse(i) {
    if (i === s.length) {
      const word = currentWord.join("");
      strings.push(word);
    } else if (
      s.charCodeAt(i) >= "A".charCodeAt(0) &&
      s.charCodeAt(i) <= "z".charCodeAt(0)
    ) {
      currentWord.push(s[i].toLowerCase());
      recurse(i + 1);
      currentWord.pop();
      currentWord.push(s[i].toUpperCase());
      recurse(i + 1);
      currentWord.pop();
    } else {
      currentWord.push(s[i]);
      recurse(i + 1);
      currentWord.pop();
    }
  }

  recurse(0);
  return strings;
}

if (process.argv.length !== 3) {
  console.warn("Usage: node solution.js {string}");
  process.exit(1);
}

const word = process.argv[2];
const list = letterCasePermutation(word);
for (const entry of list) {
  console.log(entry);
}
